### PR TITLE
gh-124160: Pass main_tstate to update_global_state_for_extension()

### PR DESCRIPTION
**Pull request by luk1337**
Otherwise it'll always return NULL if tstate != main_tstate due to _Py_IsMainInterpreter() check inside it.

Also, add a regression test that makes sure that `readline` is importable in case that triggered the crash before.

* Issue: gh-124160

### CODE REVIEW

**Comment by luk1337** [comment]
> I'll review this when I reproduce the issue later today. In the meantime, this needs a NEWS entry.

Would you be so nice to write one for me?

**Comment by luk1337** [comment]
> > I'll review this when I reproduce the issue later today. In the meantime, this needs a NEWS entry.
> 
> Would you be so nice to write one for me?

I left `Allow edits by maintainers` enabled so you can even push to my branch.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
The change is correct.  In this entire section, the main tstate is always active and should always be targeted.

Before merging this, let's add a regression test that would have caught the reported failure.

**Comment by luk1337** [comment]
> The change is correct. In this entire section, the main tstate is always active and should always be targeted.
> 
> Before merging this, let's add a regression test that would have caught the reported failure.

```diff
diff --git a/Programs/_testembed.c b/Programs/_testembed.c
index 342cc91cc58..5506ba21948 100644
--- a/Programs/_testembed.c
+++ b/Programs/_testembed.c
@@ -95,6 +95,14 @@ static void _testembed_Py_Initialize(void)
 }
 
 
+static int test_import_in_subinterpreters(void)
+{
+    _testembed_Py_InitializeFromConfig();
+    PyThreadState_Swap(Py_NewInterpreter());
+    PyRun_SimpleString("import readline"); // gh-124160
+}
+
+
 /*****************************************************
  * Test repeated initialisation and subinterpreters
  *****************************************************/
@@ -2398,6 +2406,7 @@ static struct TestCase TestCases[] = {
     {"test_repeated_init_exec", test_repeated_init_exec},
     {"test_repeated_simple_init", test_repeated_simple_init},
     {"test_forced_io_encoding", test_forced_io_encoding},
+    {"test_import_in_subinterpreters", test_import_in_subinterpreters},
     {"test_repeated_init_and_subinterpreters", test_repeated_init_and_subinterpreters},
     {"test_repeated_init_and_inittab", test_repeated_init_and_inittab},
     {"test_pre_initialization_api", test_pre_initialization_api},
```

does this look ok?

**Comment by ZeroIntensity** [comment]
Looks good, but we should be using `NULL` checks in tests.

**Comment by luk1337** [comment]
> Looks good, but we should be using `NULL` checks in tests.

There are no null checks for Py_NewInterpreter(), PyThreadState_Swap(), PyRun_SimpleString() in the whole file and this test is just crashing w/o this PR.

**Comment by ZeroIntensity** [comment]
Yeah, functions like `PyRun_SimpleString` can fail. In that case, we don't want to segfault the tests, we should just return -1 and let the error propagate.

**Comment by luk1337** [comment]
> Yeah, functions like `PyRun_SimpleString` can fail. In that case, we don't want to segfault the tests, we should just return -1 and let the error propagate.

ah indeed, I forgot about the return anyway.

**Comment by luk1337** [comment]
I have made the requested changes; please review again

**Comment by luk1337** [comment]
Added short news entry

File: Misc/NEWS.d/next/C_API/2024-09-18-18-30-16.gh-issue-124160.wFuEVx.rst
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/C_API/2024-09-18-18-30-16.gh-issue-124160.wFuEVx.rst]: Let's be a little more specific :smile:
```suggestion
Fix crash when importing modules containing state and single-phase initialization in a subinterpreter.
```

Comment (luk1337) [reply Misc/NEWS.d/next/C_API/2024-09-18-18-30-16.gh-issue-124160.wFuEVx.rst]: Applied.

File: Programs/_testembed.c
Comment (ericsnowcurrently) [inline Programs/_testembed.c:103]: This works, but only as long as the readline module fits the relevant criteria.  That's probably fine.

Ideally we would add a new test module for this dedicated case, though we could probably reuse _testsinglephase_with_reinit (in Modules/_testsinglephase.c).  If you think that would take much time, I'm fine with addressing that separately, so we can get this PR merged sooner.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM, with one caveat about the test using the readline module

File: Programs/_testembed.c
Comment (luk1337) [reply Programs/_testembed.c:103]: Would prefer if someone else figured out how to set up a more robust test case for this issue in a separate PR.

**Review by ZeroIntensity — APPROVED** [review]
I've gone through and confirmed that this fixes importing for all stateful single-phase modules. Thank you for the fix, @luk1337!

**Review by kumaraditya303 — APPROVED** [review]

**Pull request merged**
